server/src/main/java/org/apache/cassandra/sidecar/acl/AuthCache.java

```diff
@@ -0,0 +1,164 @@
+
+    // cache is null when AuthCache is disabled
+    protected volatile LoadingCache<K, V> cache;
+
+    protected AuthCache(String name,
+                        Vertx vertx,
+                        ExecutorPools executorPools,
+                        Function<K, V> loadFunction,
+                        Supplier<Map<K, V>> bulkLoadFunction,
+                        CacheConfiguration cacheConfiguration)
+    {
+        this.name = name;
+        this.vertx = vertx;
+        this.internalPool = executorPools.internal();
+        this.loadFunction = loadFunction;
+        this.bulkLoadFunction = bulkLoadFunction;
+        this.config = cacheConfiguration;
+
+        if (this.config.enabled())
+        {
+            this.cache = initCache();
+            configureSidecarServerEventListener();
+        }
+    }
+
+    /**
+     * Retrieves a value from the cache. Will call {@link LoadingCache#get(Object)} which will
+     * "load" the value if it's not present, thus populating the key. When the cache is disabled, data is fetched
+     * with loadFunction.
+     *
+     * @param k key
+     * @return The current value of {@code K} if cached or loaded.
+     * <p>
+     * See {@link LoadingCache#get(Object)} for possible exceptions.
+     */
+    public V get(K k)
+    {
+        if (!config.enabled())
+        {
+            return loadFunction.apply(k);
+        }
+        return cache.get(k);
+    }
+
+    /**
+     * Retrieves all cached entries. Will call {@link LoadingCache#asMap()} which does not trigger "load". When cache
+     * is disabled, data is fetched with bulkLoadFunction.
+     *
+     * @return a map of cached key-value pairs
+     */
+    public Map<K, V> getAll()
+    {
+        if (!config.enabled())
+        {
+            return bulkLoadFunction.get();
+        }
+        return Collections.unmodifiableMap(cache.asMap());
+    }
+
+    private LoadingCache<K, V> initCache()
+    {
+        return Caffeine.newBuilder()
+                       // setting refreshAfterWrite and expireAfterWrite to same value makes sure no stale
+                       // data is fetched after expire time
+                       .refreshAfterWrite(config.expireAfterAccessMillis(), TimeUnit.MILLISECONDS)
+                       .expireAfterWrite(config.expireAfterAccessMillis(), TimeUnit.MILLISECONDS)
+                       .maximumSize(config.maximumSize())
+                       .build(loadFunction::apply);
+    }
+
+    private void configureSidecarServerEventListener()
+    {
+        EventBus eventBus = vertx.eventBus();
+        eventBus.localConsumer(ON_ALL_CASSANDRA_CQL_READY.address(),
+                               message -> internalPool.executeBlocking(() -> {
+                                   warm(config.warmupRetries());
+                                   return null;
+                               }));
+    }
+
+    @VisibleForTesting
+    protected void warm(int availableRetries)
```

Comment:
```suggestion
    protected void warmUp(int availableRetries)
```

to avoid misreading as `warn`